### PR TITLE
Support csv format label column

### LINKED ISSUE #1715: [Need Discuss] Support simple functions on LABEL clause
Background: https://github.com/sql-machine-learning/sqlflow/issues/1643

We may have multiple target columns when training:

The task is to use `X1,X2,X3` to predict `Y1,Y2`.

The column `Y1` and `Y2` will be concatenated to a Tensor of shape [batch_size, 2] (note that typical classification jobs, the label is of shape [batchsize, 1] ). And the input features is of shape [batch_size, time_window_size, feature_size].

We can use the below approaches to achieve this:

1. write a concatenated CSV value in a single `Y` column using pure SQL before training.
1. use `LABEL CONCAT(Y1,Y2)` to do simple label transformations.

Note we don't use `LABEL Y1, Y2` since we might need to support cases using multiple training targets: https://towardsdatascience.com/regression-models-with-multiple-target-variables-8baa75aacd

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/sql-machine-learning/sqlflow/issues/1715

### CODE REVIEW

File: pkg/sql/ir_generator.go [hunk 1]
```diff
@@ -125,30 +125,30 @@ func generateTrainStmt(slct *parser.SQLFlowSelectStmt) (*ir.TrainStmt, error) {
...
 
 	if e := verifier.VerifyColumnNameAndType(tr.SQLFlowSelectStmt, pr, db); e != nil {
-		return nil, fmt.Errorf("VerifyColumnNameAndType: %v", e)
+		return nil, nil, fmt.Errorf("VerifyColumnNameAndType: %v", e)
 	}
 
 	pr.TrainClause = tr.TrainClause
 
-	return pr, nil
+	return pr, tr.SQLFlowSelectStmt, nil
```
Comment (wangkuiyi) [inline pkg/sql/ir_generator.go:151]: Does this change fix https://github.com/sql-machine-learning/sqlflow/issues/1759 ?

File: cmd/sqlflowserver/main_test.go [hunk 2]
```diff
@@ -875,6 +879,19 @@ INTO sqlflow_models.my_dnn_model_custom_functional;`
 	}
 }
 
+func CaseTrainWithCSVLabel(t *testing.T) {
```
Comment (wangkuiyi) [inline cmd/sqlflowserver/main_test.go:882]: I am a little confused -- why is the label called CSV label? CSV is a file format, right? Should we say `concat_label`?

File: pkg/step/feature/derivation.go [hunk 3]
```diff
@@ -486,6 +487,23 @@ func setBackToIR(trainStmt *ir.TrainStmt, fcMap ColumnMap, columnTargets []strin
 	}
 }
 
+// setBackToLabel update label field meta
+func setBackToLabel(trainStmt *ir.TrainStmt, fmMap FieldDescMap) {
+	// NOTE: clustering model may not specify Label
+	labelName := trainStmt.Label.GetFieldDesc()[0].Name
+	if labelName == "" {
+		return
```
Comment (wangkuiyi) [inline pkg/step/feature/derivation.go]: I see. How about moving the comment `// NOTE: clustering model doesn't have label` to the end of this line?

File: pkg/step/feature/derivation.go [hunk 4]
```diff
@@ -486,6 +487,23 @@ func setBackToIR(trainStmt *ir.TrainStmt, fcMap ColumnMap, columnTargets []strin
...
+	}
+	trainStmt.Label = &ir.NumericColumn{
+		FieldDesc: fmMap[labelName],
+	}
+	// use shape [] if label shape is [1] for Tensorflow scalar label shape should be [].
+	shape := trainStmt.Label.GetFieldDesc()[0].Shape
```
Comment (wangkuiyi) [inline pkg/step/feature/derivation.go:501]: Here I don't understand -- how comes shape? And why do we need to set the field Shape?

File: pkg/step/feature/derivation.go [hunk 5]
```diff
@@ -486,6 +487,23 @@ func setBackToIR(trainStmt *ir.TrainStmt, fcMap ColumnMap, columnTargets []strin
 	}
 }
 
+// setBackToLabel update label field meta
+func setBackToLabel(trainStmt *ir.TrainStmt, fmMap FieldDescMap) {
```
Comment (wangkuiyi) [inline pkg/step/feature/derivation.go:492]: I am not sure what does the phrase *set back* mean?

File: pkg/sql/codegen/xgboost/template_explain.go [hunk 6]
```diff
@@ -22,22 +22,22 @@ type explainFiller struct {
 	DatasetSQL           string
 	ShapSummaryParams    string
 	FeatureFieldMetaJSON string
-	LabelName            string
+	LabelJSON            string
```
Comment (wangkuiyi) [inline pkg/sql/codegen/xgboost/template_explain.go:25]: What is the relationship between label and JSON? Is that XGBoost requires JSON-encoded inputs?

**Review by wangkuiyi — COMMENTED** [review]
I am not familiar with the related source files. Just added some comments trying to better understand the code.

File: pkg/sql/ir_generator.go (hunk 1 above)
Comment (typhoonzero) [reply pkg/sql/ir_generator.go:151]: Not really, this is only a feature requested from @Echo9573 to support some time series model training, please refer to: #1715

File: cmd/sqlflowserver/main_test.go (hunk 2 above)
Comment (typhoonzero) [reply cmd/sqlflowserver/main_test.go:882]: CSV stands for "Comma Separated Values", since it's some values, I think it's OK to represent the data in the database cell.

File: pkg/sql/codegen/xgboost/template_explain.go (hunk 6 above)
Comment (typhoonzero) [reply pkg/sql/codegen/xgboost/template_explain.go:25]: It's just used to pass `ir.FieldDesc` to a Python dict object using JSON as the serialization method.

File: pkg/step/feature/derivation.go (hunk 5 above)
Comment (typhoonzero) [reply pkg/step/feature/derivation.go:492]: Indeed "set back" is not easy to understand. I was thinking, this function is setting derived column information back to the original IR structure, so it's a "set back", maybe we need a better name. I am updating this to the comment though.

File: pkg/step/feature/derivation.go (hunk 4 above)
Comment (typhoonzero) [reply pkg/step/feature/derivation.go:501]: I don't know why Tensorflow only accept shape `[]` for the label, pass `[1]` will cause an error. I'm wondering if we should pass `[None, 1]` because Tensorflow uses `None` to indicate the batch size dimension. I still need to do more tests to confirm this.

File: pkg/step/feature/derivation.go (hunk 3 above)
Comment (typhoonzero) [reply pkg/step/feature/derivation.go]: Done.

File: cmd/sqlflowserver/main_test.go (hunk 2 above)
Comment (wangkuiyi) [reply cmd/sqlflowserver/main_test.go:882]: I see. Then `CSVLabel` means `comma-separated value label`, which is ambiguous -- is it a value or a label.  Should we use the name `CaseTrainWithCommaSeparatedLabel`?

If there could be column-separated label and semi-column-separated labels in the future, we might want `ConcatenatedLabel` or `ConcatLabel` instead.

File: pkg/step/feature/derivation.go (hunk 5 above)
Comment (wangkuiyi) [reply pkg/step/feature/derivation.go:492]: Do you mean `updateLabel` or `deriveLabel` like `deriveColumn`?

**Review by wangkuiyi — APPROVED** [review]

File: cmd/sqlflowserver/main_test.go (hunk 2 above)
Comment (typhoonzero) [reply cmd/sqlflowserver/main_test.go:882]: Sounds good, will update the naming in next PR.

File: pkg/step/feature/derivation.go (hunk 5 above)
Comment (typhoonzero) [reply pkg/step/feature/derivation.go:492]: I prefer `deriveLabel` will update this in the next PR.
